cmd/shim/secret_scrub.go

```diff
@@ -0,0 +1,99 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/fs"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/dagger/dagger/core"
+)
+
+type SecretScrubWriter struct {
+	mu           sync.Mutex
+	w            io.Writer
+	secretValues []string
+	fs           fs.FS
+}
+
+func (w *SecretScrubWriter) Write(b []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	s := string(b)
+	for _, secret := range w.secretValues {
+		if secret == "" {
+			continue
+		}
+		// FIXME: I think we can do better
+		s = strings.ReplaceAll(s, secret, "***")
+	}
+
+	_, err := w.w.Write([]byte(s))
+	if err != nil {
+		return -1, err
+	}
+
+	return len(b), err
+}
+
+// NewSecretScrubWriter replaces known secrets by "***".
+// The value of the secrets referenced in secretsToScrub are loaded either
+// from env or from the fs.
+func NewSecretScrubWriter(w io.Writer, fsys fs.FS, env []string, secretsToScrub core.SecretToScrubInfo) (io.Writer, error) {
+	secrets := loadSecretsToScrubFromEnv(env, secretsToScrub.Envs)
+
+	fileSecrets, err := loadSecretsToScrubFromFiles(fsys, secretsToScrub.Files)
+	if err != nil {
+		return nil, fmt.Errorf("could not load secrets from file: %w", err)
+	}
+	secrets = append(secrets, fileSecrets...)
+
+	return &SecretScrubWriter{
+		fs:           fsys,
+		w:            w,
+		secretValues: secrets,
+	}, nil
+}
+
+func loadSecretsToScrubFromEnv(env []string, secretsToScrub []string) []string {
+	secrets := []string{}
+
+	for _, envKV := range env {
+		envName, envValue, ok := strings.Cut(envKV, "=")
+		// no env value for this secret
+		if !ok {
+			continue
+		}
+
+		for _, envToScrub := range secretsToScrub {
+			if envName == envToScrub {
+				secrets = append(secrets, envValue)
+			}
+		}
+	}
+
+	return secrets
+}
+
+func loadSecretsToScrubFromFiles(fsys fs.FS, secretFilePathsToScrub []string) ([]string, error) {
+	secrets := make([]string, 0, len(secretFilePathsToScrub))
+
+	for _, fileToScrub := range secretFilePathsToScrub {
+		absFileToScrub, err := filepath.Abs(fileToScrub)
+		if err != nil {
+			return secrets, err
+		}
```

Comment:
nit: It's a little unfortunate that this defeats the point of using `fs.FS` now that it's coupled to the host filesystem (by my own suggestion, oops). I don't think the tests reveal that at the moment because all the paths in the tests are already absolute.

We could maybe pass in a workdir in addition to the `fs.FS` and just `path.Join` it if `!path.IsAbs(fileToScrub)`.

```suggestion
		absFileToScrub := fileToScrub
		if !path.IsAbs(fileToScrub) {
			absFileToScrub = path.Join(workDir, fileToScrub)
		}
```

That way we can have a test for relative paths too. In the real code we'd just call `os.Getwd()` and pass it in.

Reply:
I think that the secret file mount path will be absolute anyway: https://github.com/dagger/dagger/blob/b33af8856b56ac59d57983aaf58050c46ec3b2e9/core/container.go#L561

I guess, without that, it would be impossible to know where to find files anyway as the exec would happen at the end, and the workdir of the exec could be different from the time when the files were mounted.

But I agree that adding the dir path here would make it more testable.

Reply:
done.

Reply:
Oh, you're right, we don't need all this relative/absolute complexity at all. Sorry!